theano/tensor/nnet/bn.py

```diff
@@ -0,0 +1,70 @@
+import theano
+from theano.scalar import Composite
+from theano.scalar import add, sub, true_div, mul
+
+
+class BNComposite(Composite):
+
+    def __init__(self, dtype):
+        x = theano.scalar.Scalar(dtype=dtype).make_variable()
+        mean = theano.scalar.Scalar(dtype=dtype).make_variable()
+        std = theano.scalar.Scalar(dtype=dtype).make_variable()
+        gamma = theano.scalar.Scalar(dtype=dtype).make_variable()
+        beta = theano.scalar.Scalar(dtype=dtype).make_variable()
+        o = add(mul(true_div(sub(x, mean), std), gamma), beta)
+        inputs = [x, mean, std, gamma, beta]
+        outputs = [o]
+        super(BNComposite, self).__init__(inputs, outputs)
+
+    def grad(self, inps, grads):
+        x, mean, std, gamma, beta = inps
+        top, = grads
+        dx = (top * gamma) / std
+        dmean = -(top * gamma) / std
+        dstd = -(top * gamma * (x - mean)) / (std * std)
+        dgamma = top * (x - mean) / std
+        return [dx, dmean, dstd, dgamma, top]
+
+
+def batch_normalization(inputs, gamma, beta, mean, std,
+                        mode='low_mem'):
+    """
+    This function will build the symbolic graph for applying batch normalization
+    to a set of activations.
+    Work also on GPU
+
+    Parameters
+    ----------
+    inputs : symbolic tensor
+        Mini-batch of activations
+    gamma: symbolic tensor
+        BN scale parameter, must be of same dimensionality as
+        inputs and broadcastable against it
+    beta: symbolic tensor
+        BN shift parameter, must be of same dimensionality as
+        inputs and broadcastable against it
+    mean: symbolic tensor
+        inputs means, must be of same dimensionality as
+        inputs and broadcastable against it
+    std: symbolic tensor
+        inputs standard deviation, must be of same dimensionality as
+        inputs and broadcastable against it
+    mode: 'low_mem' or 'high_mem'
+        Specify which batch_normalization implementation that will be
+        used.
+        As no intermediate representations are stored for the back-propagation,
+        'low_mem' implementation lower the memory usage, however,
+        it is 5-10% slower than 'high_mem' implementation. Note that 5-10% computation
+        time difference compare the batch_normalization operation only, time difference
+        between implementation is likely to be less important on the full model fprop/bprop.
+    """
+    if mode == 'low_mem':
+        elm_bn = theano.tensor.elemwise.Elemwise(scalar_op=BNComposite(dtype=inputs.dtype))
+        rval = elm_bn(inputs, mean, std, gamma, beta)
+    elif mode == 'high_mem':
+        rval = (inputs - mean) / std
+        rval = rval * gamma + beta
```

Comment:
I think you are right. But why you say too late? This is just changing the
order or a multiplication and should give the same value to eps. We do
optimization that change the eps, so I don't see a problem to change it
there. Here is the PR that do this:

https://github.com/Theano/Theano/pull/3536

thanks

On Mon, Oct 19, 2015 at 2:37 PM, Jan Schlüter <notifications@github.com>
wrote:

> In theano/tensor/nnet/bn.py
> <https://github.com/Theano/Theano/pull/3410#discussion_r42409021>:
>
> > +        inputs and broadcastable against it
> > +    mode: 'low_mem' or 'high_mem'
> > +        Specify which batch_normalization implementation that will be
> > +        used.
> > +        As no intermediate representations are stored for the back-propagation,
> > +        'low_mem' implementation lower the memory usage, however,
> > +        it is 5-10% slower than 'high_mem' implementation. Note that 5-10% computation
> > +        time difference compare the batch_normalization operation only, time difference
> > +        between implementation is likely to be less important on the full model fprop/bprop.
> > +    """
> > +    if mode == 'low_mem':
> > +        elm_bn = theano.tensor.elemwise.Elemwise(scalar_op=BNComposite(dtype=inputs.dtype))
> > +        rval = elm_bn(inputs, mean, std, gamma, beta)
> > +    elif mode == 'high_mem':
> > +        rval = (inputs - mean) / std
> > +        rval = rval * gamma + beta
>
> Noticed that too late, but wouldn't rval = (inputs - mean) * (gamma /
> std) + beta usually be a bit more efficient? std and gamma are usually
> smaller than inputs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3410/files#r42409021>.
>